[xpu][test][1/N] Enable tests of test_nn.py on Intel GPU - instantiate TestNN with instantiate_device_type_tests

daisyden opened the pull request:
For https://github.com/pytorch/pytorch/issues/114850, we will port aten unit tests to Intel GPU. This PR will work on test/test_nn.py TEST_NN class for single GPU test only. We could enable Intel GPU with following methods and try the best to keep the original code styles:

1. Use torch.accelerator to extend cude specific test to XPU.
2. Added skipIfXPU decorator for cases with known issues on Intel GPU
3. Enabled 'xpu' for some test pathes

daisyden committed:
run TestNN with instantiate_device_type_tests

guangyey commented on test/test_nn.py:
```diff
@@ -227,12 +231,13 @@ def test_to(self):
-            for cuda in ['cuda', 'cuda:0' if torch.cuda.device_count() == 1 else 'cuda:1']:
-                m2 = m.cuda(device=cuda)
-                self.assertIs(m2, m2.to(cuda))
+        if torch.cuda.is_available() or torch.xpu.is_available():
```
```suggestion
        if TEST_GPU :
```

guangyey commented on test/test_nn.py:
```diff
@@ -2784,15 +2797,15 @@ def test_CTCLoss_critical_target_len(self):
     def test_CTCLoss_zero_lengths(self):
         devices = ['cpu']
-        devices += ['cuda'] if TEST_CUDA else []
+        devices += [device_type] if TEST_CUDA or TEST_XPU else []
```
```suggestion
        devices += [device_type] if TEST_GPU else []
```

guangyey commented on test/test_nn.py:
```diff
@@ -2814,16 +2827,16 @@ def test_CTCLoss_zero_lengths(self):
             self.assertTrue((inp.grad == 0).all().item())
 
-    @unittest.skipIf(not TEST_CUDA, 'CUDA not available')
+    @unittest.skipIf(not TEST_CUDA and not TEST_XPU, 'CUDA and XPU not available')
```
```suggestion
    @unittest.skipIf(not TEST_GPU, 'CUDA and XPU not available')
```

guangyey commented on test/test_nn.py:
```diff
@@ -5008,17 +5021,17 @@ def helper(self, dtype):
-    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
-    @unittest.skipIf(not TEST_CUDNN, "needs cudnn")
+    @unittest.skipIf(not TEST_CUDA and not TEST_XPU, "CUDA and XPU unavailable")
+    @unittest.skipIf(not TEST_CUDNN and not TEST_XPU, "needs cudnn or xpu")
```
Here don't need `TEST_XPU`, right

guangyey commented on test/test_nn.py:
```diff
@@ -2814,16 +2827,16 @@ def test_CTCLoss_zero_lengths(self):
             self.assertTrue((inp.grad == 0).all().item())
 
-    @unittest.skipIf(not TEST_CUDA, 'CUDA not available')
+    @unittest.skipIf(not TEST_GPU, 'CUDA and XPU not available')
```
```suggestion
    @unittest.skipIf(not TEST_GPU, 'The current accelerator is not available')
```

guangyey commented on test/test_nn.py:
```diff
@@ -5199,7 +5212,14 @@ def test_batchnorm(self, dims, mode, memory_format, ref_backend, mixed, dtype):
-                self.skipTest("Failed on CUDA")
+                self.skipTest(f"Failed on {device_type.upper()}")
+
+        if torch.version.xpu:
```
```suggestion
        if TEST_XPU:
```

guangyey reviewed (dismissed):
lgtm.

malfet reviewed (changes requested):
Is it possible to split this change into few smaller ones? For example, change to `common_device_types` and `test_namedtensor.py` looks totally fine and majority of the changes to `test_nn` are also OK, but multigpu changes seems a bit risky

Perhaps let's do single GPU enablement in 1 PR and than multi-GPU in another

daisyden commented on test/test_nn.py:
```diff
@@ -75,14 +75,18 @@
     import numpy as np
 
+TEST_GPU = torch.cuda.is_available() or torch.xpu.is_available()
+TEST_MULTIGPU = TEST_GPU and torch.accelerator.device_count() >= 2
```
Can I add TEST_GPU and TEST_MULTIGPU definition in common_utils.py?

daisyden replied:
Hi, @malfet, I removed the code updates for multi-gpu from this PR and will create another PR for them. I also moved the TEST_GPU definition to common_utils. Please further review this PR.  Thanks.

daisyden commented on test/test_nn.py:
```diff
@@ -4388,9 +4401,9 @@ def test_RNN_input_size_zero(self):
-                if device == 'cuda':
-                    rnn.cuda()
-                    input = input.cuda()
+                if device == 'cuda' or device == 'xpu':
```
Done, this PR is now only for single gpu.

EikanWang commented on test/test_nn.py:
```diff
@@ -3199,8 +3211,8 @@ def get_a_test_layer(use_cuda, activation, batch_first=False):
             activation = F.relu
-            use_cuda = torch.cuda.is_available()
-            device = torch.device("cuda" if use_cuda else "cpu")
+            use_cuda = TEST_GPU
```
```suggestion
```

EikanWang commented on test/test_nn.py:
```diff
@@ -3424,8 +3436,8 @@ def perm_fn(x):
+            use_cuda = TEST_GPU
+            device = torch.device(device_type if use_cuda else "cpu")
 
             decoder_layer = get_a_test_layer(use_cuda=use_cuda, activation=activation,
```
```suggestion
            decoder_layer = get_a_test_layer(activation=activation,
```

malfet reviewed (dismissed):
This PR achieves the goal of enabling tests for XPU, but imo it's goes contrary to the goal of generalizing the tests/using torch.accelerator API. I.e. why those tests should not be running on some PrivateUse1 devices or say MPS?


I.e. perhaps it could be split into two PRs:
 - one that adds `device` argument to the tests and decorates CUDA only tests with `@onlyAccelerator` and 2nd one that instantiates subset of tests to GPU

daisyden commented on test/test_nn.py:
```diff
@@ -2318,30 +2323,32 @@ def forward(self, x):
 
+    @skipIfXpu
     @unittest.skipIf(not TEST_MULTIGPU, "multi-GPU not supported")
     def test_broadcast_not_requiring_grad(self):
```
@malfet  Thanks for the suggestion, I updated this PR to instantiate TestNN with instantiate_device_type_tests(). Please help review the code.  I will enable XPU test in another PR when the first PR is completed.

daisyden commented on test/test_nn.py:
```diff
@@ -4286,8 +4298,8 @@ def check_weight_norm(m, name):
             m = torch.nn.utils.weight_norm(m, name=name)
 
             # moves to CUDA
```
cudnn test only, so keep the code unchanged.

mikaylagawarecki commented on test/test_nn.py:
```diff
@@ -4418,14 +4421,14 @@ def test_error_RNN_seq_len_zero(self):
-    def test_RNN_input_size_zero(self):
+    def test_RNN_input_size_zero(self, device):
         for module in (nn.RNN, nn.LSTM, nn.GRU):
             for device in get_all_device_types():
```
the test no longer needs to iterate over `get_all_device_types` when it has been instantiate with instantiate_device_type_tests

mikaylagawarecki commented on test/test_nn.py:
```diff
@@ -2815,15 +2821,15 @@ def test_CTCLoss_critical_target_len(self):
 
-    def test_CTCLoss_zero_lengths(self):
+    def test_CTCLoss_zero_lengths(self, device):
         devices = ['cpu']
```
you can delete `devices`, there's no need to loop over all devices if we already use `instantiate_device_type_tests`

mikaylagawarecki commented on test/test_nn.py:
```diff
@@ -5929,11 +5934,11 @@ def test_affine_grid_backward_cl_cf_consistency(self, device, nd):
         # so we test both cases.
-        def test(N, C, H, W, mode, padding_mode, align_corners, input_requires_grad):
-            def test_shape(N, C, IH, IW, H, W, mode, padding_mode, align_corners):
+        def test(N, C, H, W, mode, padding_mode, align_corners, input_requires_grad, device='cuda'):
```
nit: perhaps no need for default here 
```suggestion
        def test(N, C, H, W, mode, padding_mode, align_corners, input_requires_grad, device):
```

mikaylagawarecki commented on test/test_nn.py:
```diff
@@ -14231,7 +14234,8 @@ def __init__(self) -> None:
 
 instantiate_device_type_tests(TestNNDeviceType, globals(), allow_mps=True)
-instantiate_parametrized_tests(TestNN)
+instantiate_device_type_tests(TestNN, globals())
```
As a side effect, does this make all the tests that are not passed the `device` arg run the exact same code twice? Once for TestNNCPU and once for TestNN{Device}?

mikaylagawarecki reviewed (commented):
Thanks for your work on this PR! There are 3 high level comments that I have 

- Several of the tests in this PR have an additional internal loop that loops over a `device_list` and runs code on each device, despite having instantiated the tests for each device with `instantiate_device_type_tests`. The internal loop can be removed
- See [this comment](https://github.com/pytorch/pytorch/pull/166396/changes#r2891973959) re `HAS_ACCELERATOR` within the test body 
- Some of the tests marked as onlyAccelerator seem to be intended to be cudnn tests per their name.

mikaylagawarecki commented:
@claude This PR intends to make the `TestNN` class device-generic. Review the changes in this PR, paying particular attention to 1. whether any existing test functionality was changed 2. whether the contents of the modified tests still make sense given that the test class was migrated to use `instantiate_device_type_tests` 3. whether the tests marked onlyAccelerator should run on all accelerators

daisyden commented:
> Thanks for your work on this PR! There are 3 high level comments that I have
> 
> * Several of the tests in this PR have an additional internal loop that loops over a `device_list` and runs code on each device, despite having instantiated the tests for each device with `instantiate_device_type_tests`. The internal loop can be removed
> * See [this comment](https://github.com/pytorch/pytorch/pull/166396/changes#r2891973959) re `HAS_ACCELERATOR` within the test body
> * Some of the tests marked as onlyAccelerator seem to be intended to be cudnn tests per their name.

@mikaylagawarecki  thanks for your valuable suggestions. Basically I agree with you on the three points and I will update my code.

daisyden committed:
Generalize the TestNN further: 1. Replace TEST_CUDA with 'ACCELERATOR_TYPE == device', 2. remove device_() loop as the test for the device is instantiated with instantiate_device_type_tests, 3. remove unnessesary default value 4. skip the failed cases on MPS backend and CPU dynamic_wrapper test

A testing commit to check whehter CI failure in test_stream_event() is introduced by common code change

Fix ACCELERATOR_TYPE definition with LazyVal, updated onlyAccelerator
implementation.

Revert "A testing commit to check whehter CI failure in test_stream_event() is introduced by common code change"
Disable cpu dynamo_wrapped test due to known issue

This reverts commit 6d4e76ff0659d7b379a41e43e34a964840252aa1.
skip cpu test when PYTORCH_TEST_WITH_DYNAMO is True due to #177119

make largeTensorTest general for all devices when onlyAccelerator is decorated, remove unnessesary device list furher, remove onlyAccelerator for CUDNN test
comment PYTORCH_TEST_WITH_DYNAMO checker to reproduce cpu issue 177119

guangyey commented on test/test_nn.py:
```diff
@@ -5227,14 +5226,17 @@ def test_batchnorm_buffer_update_when_stats_are_not_tracked(self):
-                                        "test_batchnorm_3D_train_NCHW_vs_native_mixed_float16"):
-                self.skipTest("Failed on CUDA")
+    def test_batchnorm(self, device, dims, mode, memory_format, ref_backend, mixed, dtype):
+        if ACCELERATOR_TYPE == device:
```
This code change changes ROCM's behavior.

daisyden replied:
updated

daisyden commented on test/test_nn.py:
```diff
@@ -4333,7 +4345,7 @@ def test_RNN_dropout(self):
                 for cuda in (True, False):
                     rnn = nn.RNN(10, 1000, 2, bias=False, dropout=p, nonlinearity='relu')
                     if cuda:
```
Cudnn test, we could not change it in this PR.

daisyden commented on test/test_nn.py:
```diff
@@ -4444,8 +4456,8 @@ def test_RNN_change_dropout(self):
             rnn = nn.RNN(100, 100, 2, dropout=0, nonlinearity='relu')
             input = torch.rand(3, 2, 100)
             if cuda:
```
cudann test, will not update it in the PR.

daisyden commented on test/test_nn.py:
I will simplify test_cosine_similarity_mixed_precision.

mikaylagawarecki commented on test/test_nn.py:
```diff
@@ -6443,13 +6396,13 @@ def test_shape(N, C, ID, IH, IW, D, H, W, mode, padding_mode, align_corners):
-        if TEST_CUDA:
-            device_list.append('cuda')
+        if ACCELERATOR_TYPE == device:
+            device_list.append(device)
```
why does this test still have the loop over `device_list`?

mikaylagawarecki reviewed (commented):
Thanks for your effort on this, I know it is not easy to go make the entire class device-generic!

I went over the PR manually and I think most of my comments ended up matching the claude comments.  I think this is pretty close, just a couple more small comments

daisyden commented on test/test_nn.py:
```diff
@@ -14277,7 +14214,14 @@ def __init__(self) -> None:
+    device_list = ('cuda', 'cpu')
+else:
+    device_list = ('cuda',)
+instantiate_device_type_tests(TestNN, globals(), only_for=tuple(device_list))
```
The skipIfMPS in this PR are for TestNNDeviceType which is instantiate with instantiate_device_type_tests(TestNNDeviceType, globals(), allow_mps=True).

daisyden commented on test/test_nn.py:
```diff
@@ -14277,7 +14214,14 @@ def __init__(self) -> None:
+
+# https://github.com/pytorch/pytorch/issues/177119
+if os.environ.get('PYTORCH_TEST_WITH_DYNAMO', '0') == '0':
+    device_list = ('cuda', 'cpu')
```
Can I use except_for? 
```
if os.environ.get('PYTORCH_TEST_WITH_DYNAMO', '0') == '1':
    device_list = ('cpu', )

instantiate_device_type_tests(TestNN, globals(), except_for=device_list)
```

daisyden commented:
@claude Could you do another pass on this PR, taking into account previous review comments. Make sure that the functionality being tested was not changed.

mikaylagawarecki reviewed (approved):
Thank you! 

For my understanding -- the main change in this PR was to instantiate `TestNN` with `instantiate_device_type_tests`. `TestNN` now no longer has `@onlyCUDA`. It seems like this PR also migrated some (but not all) `@onlyCUDA` in `TestNNDeviceType` to `@onlyAccelerator`. Will there be a follow-up PR that removes more `@onlyCUDA` from `TestNNDeviceType` (or moves them into a separate class if they are truly cuda-specific)?

malfet reviewed (changes requested):
Two comments:
 - Test plan for changes like that should include links to some runs before and after the change to guarantee there are no significant changes in the number of tests run before/after change (as this have a potential of skipping lots of those for no good reasion
 - Never use `skipIf` decorator unless it hard crashes, instead use `extectedFailureIf`

mikaylagawarecki commented:
@daisyden I am sorry it is my bad that some of my comments led to more changes being put into this PR. Perhaps we can break it up like

1. Instantiate TestNN with instantiate_device_type_tests, change 'cuda' → device                                                                                                         
2. Replace unittest.skipIf(TEST_CUDA) / onlyCUDA with onlyAccelerator as appropriate
3. Remove unnecessary internal loops over device

 @malfet let me know if this does not sound reasonable

daisyden committed:
add onlyAccelerator to facilitate onlyCUDA test generalization to multiple devices

daisyden committed:
run TestNN with instantiate_device_type_tests, rename the TestNN.test_<name> under dynamo_skips and compiled_autograd_skips to TestCPU.test_<name>_cpu
